pkg/server/plugin/upstreamca/awscertificatemanager/awscertificatemanager.go

```diff
@@ -0,0 +1,223 @@
+func (*CertificateManagerPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+// SubmitCSR submits the CSR to ACM PCA to be signed by the certificate authority
+func (m *CertificateManagerPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCSRRequest) (*upstreamca.SubmitCSRResponse, error) {
+	csrBuf := new(bytes.Buffer)
+	err := pem.Encode(csrBuf, &pem.Block{
+		Type:  csrRequestType,
+		Bytes: request.Csr,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Have ACM sign the certificate
+	m.log.Info("Submitting CSR to ACM.", "signingAlgorithm", m.signingAlgorithm)
+	validityPeriod := time.Second * time.Duration(request.PreferredTtl)
+	issueResponse, err := m.certificateManagerClient.IssueCertificateWithContext(ctx, &acmpca.IssueCertificateInput{
+		CertificateAuthorityArn: aws.String(m.certificateAuthorityArn),
+		SigningAlgorithm:        aws.String(m.signingAlgorithm),
+		Csr:                     csrBuf.Bytes(),
+		TemplateArn:             aws.String(m.caSigningTemplateArn),
+		Validity: &acmpca.Validity{
+			Type:  aws.String(acmpca.ValidityPeriodTypeAbsolute),
+			Value: aws.Int64(m.hooks.clock.Now().Add(validityPeriod).Unix()),
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Using the output of the `IssueCertificate` call, poll ACM until
+	// the certificate has been issued
+	certificateArn := issueResponse.CertificateArn
+
+	m.log.Info("Waiting for issuance from ACM.", "certificateARN", aws.StringValue(certificateArn))
+	getCertificateInput := &acmpca.GetCertificateInput{
+		CertificateAuthorityArn: aws.String(m.certificateAuthorityArn),
+		CertificateArn:          certificateArn,
+	}
+	err = m.certificateManagerClient.WaitUntilCertificateIssuedWithContext(ctx, getCertificateInput)
+	if err != nil {
+		return nil, err
+	}
+	m.log.Info("Certificate issued.", "certificateARN", aws.StringValue(certificateArn))
+
+	// Finally get the certificate contents
+	m.log.Info("Retrieving certificate and bundle from ACM.", "certificateARN", aws.StringValue(certificateArn))
+	getResponse, err := m.certificateManagerClient.GetCertificateWithContext(ctx, getCertificateInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the cert from the response
+	cert, err := pemutil.ParseCertificate([]byte(aws.StringValue(getResponse.Certificate)))
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the chain from the response
+	certChain, err := pemutil.ParseCertificates([]byte(aws.StringValue(getResponse.CertificateChain)))
+	if err != nil {
+		return nil, err
+	}
+	m.log.Info("Certificate and bundle received.", "certificateARN", aws.StringValue(certificateArn))
+
+	// Assemble the bundle
+	var bundle []byte
+	for _, cert := range certChain {
+		bundle = append(bundle[:], cert.Raw[:]...)
+	}
+
+	return &upstreamca.SubmitCSRResponse{
+		SignedCertificate: &upstreamca.SignedCertificate{
+			CertChain: cert.Raw,
+			Bundle:    bundle,
```

Comment:
> Not sure if I'm grok'ing this correctly... CertChain should be the new intermediate CA cert + higher order intermediates, but not the root. Bundle should contain the root(s)... but here it looks like we are setting CertChain to the new intermediate CA cert, and bundle to the union of the higher order intermediates?

Yeah, I think this is a mistake on my part, but you are correct the the ACM API returns all signers up the chain, including the root. Here's an example of a certificate chain returned for a hierarchy like `root -> intermediate -> spire server intermediate`:

```bash
$ openssl crl2pkcs7 -nocrl -certfile test-chain.pem | openssl pkcs7 -print_certs -noout

subject=C = US, ST = Washington, L = Seattle, O = SPIFFE Test, OU = SPIRE, CN = example.intermediate
issuer=C = US, ST = Washington, L = Seattle, O = SPIFFE Test, OU = SPIRE, CN = example.root

subject=C = US, ST = Washington, L = Seattle, O = SPIFFE Test, OU = SPIRE, CN = example.root
issuer=C = US, ST = Washington, L = Seattle, O = SPIFFE Test, OU = SPIRE, CN = example.root
```

> Maybe it is not possible to have a non-root (i.e. non-self-signed) stored in ACM PCA?

You can have a CA hierarchy of up to 5 levels in ACM PCA. See [Designing a CA hierarchy](https://docs.aws.amazon.com/acm-pca/latest/userguide/ca-hierarchy.html) for some detail on this.

I assume there are no constraints with SPIFFE/SPIRE on the upstream CA hierarchy (nothing mentioned in the [XVID specification](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md#32-signing-certificates)), and we'd want the bundle to include everything above the server's intermediate CA? 

That could be reliably assembled with a `GetCertificateAuthorityCertificate` call on the configured upstream CA, otherwise it would require parsing the returned chain from a `GetCertificate` chain.

Reply:
Correct, no limitation or constraint on path length in SPIRE

Here in the `cert_chain` field of the plugin response `upstreamca.SignedCertificate`, we expect both the newly minted cert (SPIRE's cert) and the rest of the chain, excluding the root. In the `bundle` field, we expect one or more root certificates.

Taking your above configuration as an example, the value (and order) of cert_chain should be `spire server intermediate, intermediate` and the value of bundle should be `root`.

Putting all of the intermediates in the bundle (as the current code does) causes SPIRE to not include them in SVID chains that it issues to workloads. Instead, they are distributed as part of the bundle. The impact there is that if you have a workload serving TLS, it will serve a partial chain, and if the caller doesn't have the root _plus_ the rest of the intermediates, validation will fail.

Reply:
> and if the caller doesn't have the root plus the rest of the intermediates, validation will fail.

I guess I should mention that this may be the case for callers not using SPIRE (since if it was using SPIRE, it would have the intermediates). People may bake only the root into images etc

Reply:
Chain and bundle are now parsed appropriately.